### PR TITLE
Issue #4682: add scripts for checking idea inspections

### PR DESCRIPTION
**Pull request by Nimfadora**
Issue #4682

### CODE REVIEW

File: idea_inspection.bat [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	echo IntelliJ IDEA was not found in path
+)
+
+::Ask user for input
+SET /P IDEA_LOCATION=Enter a location of IntelliJ IDEA:
+SET IDEA_LOCATION=%IDEA_LOCATION%\bin\inspect.bat
+
+
+:run
+%IDEA_LOCATION% %PROJECT_DIR% %INSPECTIONS_PATH% %RESULTS_DIR% -%NOISE_LVL%
```
Comment (rnveach) [inline idea_inspection.bat]: Please add newline to end of file.

**Comment by Nimfadora** [comment]
This script has no ability to configure the scope of inspections yet. Apparently, this is imposible by [idea inspections console tool](https://www.jetbrains.com/help/idea/running-inspections-offline.html) .
I have found the only one workaround to do it, but it is quite tricky - https://github.com/bentolor/idea-cli-inspector/blob/master/ideainspect.groovy#L249 .

So, I think, would it be a better choice to utilize that groovy script instead of rewriting its logic for setting scope via properties?

**Comment by romani** [comment]
> This script has no ability to configure the scope of inspections yet.

scope file has to be copied to IDEA folder , see instructions at http://checkstyle.sourceforge.net/idea.html#Inspections "To add custom scope copy file....." please add this to bat file as instruction on how to use.

**Comment by Nimfadora** [comment]
@romani 

> Please test if that works.
Nope, it doesn't. It just inspects the whole project.
And I've tested a solution in the groovy script I mentioned before - it doesn't work also (idea 2017.1.5).

Looks like there is no way to provide a scope... At least, I did not find any.

**Comment by Nimfadora** [comment]
@romani by the way ...

> scope file has to be copied to IDEA folder , see instructions at http://checkstyle.sourceforge.net/idea.html#Inspections "To add custom scope copy file....."

This approach of importing scope no longer work for idea 2017 (I've tested 2017.1.5 and 2017.2.1)

**Comment by romani** [comment]
@Nimfadora , strange , it works fine always for me, even for 2017.1.4 that is currently I use. Did you relaunch IDEA after this folder/file creation ? Is it for both Community edition and Enterprise ?

@MEZk , @Vladlis , can you recheck that it works for your version of IDEA ?

**Comment by Nimfadora** [comment]
@romani I have tested on windows, as this is a windows skript we are trying to write. As far as I concerned you have Linux.

**Comment by romani** [comment]
@Nimfadora ,

> This script has no ability to configure the scope of inspections yet.

settings from TeamCity
image
As you can notice it is duplicate content of `Checkstyle_Inspection_Scope.xml` in `Include / exclude patterns:`

**Comment by romani** [comment]
we released 8.2 version, please rebase all your PRs to our latest master to avoid CI problems

**Comment by Nimfadora** [comment]
@romani ok, this was helpful. ` -Didea.exclude.patterns=....` really works. But this approach makes all more difficult.

The problem is that this flag cannot be used neither with `inspect.bat` or `idea.bat` (`inspect.bat` ends with call to  `idea.bat`)

Just for convinience - here is source of my `idea.bat`

```
::@ECHO OFF

::----------------------------------------------------------------------
:: IntelliJ IDEA startup script.
::----------------------------------------------------------------------

:: ---------------------------------------------------------------------
:: Ensure IDE_HOME points to the directory where the IDE is installed.
:: ---------------------------------------------------------------------
SET IDE_BIN_DIR=%~dp0
SET IDE_HOME=%IDE_BIN_DIR%\..

:: ---------------------------------------------------------------------
:: Locate a JDK installation directory which will be used to run the IDE.
:: Try (in order): IDEA_JDK, idea%BITS%.exe.jdk, ..\jre, JDK_HOME, JAVA_HOME.
:: ---------------------------------------------------------------------
SET JDK=

IF EXIST "%IDEA_JDK%" SET JDK=%IDEA_JDK%
IF EXIST "%JDK%" GOTO check

SET BITS=64
SET USER_JDK64_FILE=%USERPROFILE%\.IdeaIC2017.1\config\idea%BITS%.exe.jdk
SET BITS=
SET USER_JDK_FILE=%USERPROFILE%\.IdeaIC2017.1\config\idea%BITS%.exe.jdk
IF EXIST "%USER_JDK64_FILE%" (
  SET /P JDK=<%USER_JDK64_FILE%
) ELSE (
  IF EXIST "%USER_JDK_FILE%" SET /P JDK=<%USER_JDK_FILE%
)
IF NOT "%JDK%" == "" (
  IF NOT EXIST "%JDK%" SET JDK="%IDE_HOME%\%JDK%"
  IF EXIST "%JDK%" GOTO check
)

IF EXIST "%IDE_HOME%\jre64" SET JDK=%IDE_HOME%\jre64
IF EXIST "%JDK%" GOTO check

IF EXIST "%IDE_HOME%\jre32" SET JDK=%IDE_HOME%\jre32
IF EXIST "%JDK%" GOTO check

IF EXIST "%JDK_HOME%" SET JDK=%JDK_HOME%
IF EXIST "%JDK%" GOTO check

IF EXIST "%JAVA_HOME%" SET JDK=%JAVA_HOME%

:check
SET JAVA_EXE=%JDK%\bin\java.exe
IF NOT EXIST "%JAVA_EXE%" SET JAVA_EXE=%JDK%\jre\bin\java.exe
IF NOT EXIST "%JAVA_EXE%" (
  ECHO ERROR: cannot start IntelliJ IDEA.
  ECHO No JDK found. Please validate either IDEA_JDK, JDK_HOME or JAVA_HOME points to valid JDK installation.
  ECHO
  EXIT /B
)

SET JRE=%JDK%
IF EXIST "%JRE%\jre" SET JRE=%JDK%\jre
IF EXIST "%JRE%\lib\amd64" SET BITS=64

:: ---------------------------------------------------------------------
:: Collect JVM options and properties.
:: ---------------------------------------------------------------------
IF NOT "%IDEA_PROPERTIES%" == "" SET IDE_PROPERTIES_PROPERTY="-Didea.properties.file=%IDEA_PROPERTIES%"

SET USER_VM_OPTIONS_FILE=%USERPROFILE%\.IdeaIC2017.1\idea%BITS%.exe.vmoptions
SET VM_OPTIONS_FILE=%IDE_BIN_DIR%\idea%BITS%.exe.vmoptions
IF EXIST "%IDE_BIN_DIR%\win\idea%BITS%.exe.vmoptions" SET VM_OPTIONS_FILE=%IDE_BIN_DIR%\win\idea%BITS%.exe.vmoptions
IF EXIST %USER_VM_OPTIONS_FILE% SET VM_OPTIONS_FILE=%USER_VM_OPTIONS_FILE%
IF NOT "%IDEA_VM_OPTIONS%" == "" SET VM_OPTIONS_FILE=%IDEA_VM_OPTIONS%

SET ACC=
FOR /F "usebackq delims=" %%i IN ("%VM_OPTIONS_FILE%") DO CALL "%IDE_BIN_DIR%\append.bat" "%%i"
IF EXIST "%VM_OPTIONS_FILE%" SET ACC=%ACC% -Djb.vmOptionsFile="%VM_OPTIONS_FILE%"

SET COMMON_JVM_ARGS="-XX:ErrorFile=%USERPROFILE%\java_error_in_IDEA_%%p.log" "-XX:HeapDumpPath=%USERPROFILE%\java_error_in_IDEA.hprof" "-Xbootclasspath/a:%IDE_HOME%/lib/boot.jar" -Didea.paths.selector=IdeaIC2017.1 %IDE_PROPERTIES_PROPERTY%
SET IDE_JVM_ARGS=-Didea.platform.prefix=Idea -Didea.jre.check=true
SET ALL_JVM_ARGS=%ACC% %COMMON_JVM_ARGS% %IDE_JVM_ARGS% 
SET CLASS_PATH=%IDE_HOME%\lib\bootstrap.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\extensions.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\util.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\jdom.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\log4j.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\trove4j.jar
SET CLASS_PATH=%CLASS_PATH%;%IDE_HOME%\lib\jna.jar
SET CLASS_PATH=%CLASS_PATH%;%JDK%\lib\tools.jar
IF NOT "%IDEA_CLASS_PATH%" == "" SET CLASS_PATH=%CLASS_PATH%;%IDEA_CLASS_PATH%

:: ---------------------------------------------------------------------
:: Run the IDE.
:: ---------------------------------------------------------------------
SET OLD_PATH=%PATH%
SET PATH=%IDE_BIN_DIR%;%PATH%

"%JAVA_EXE%" %ALL_JVM_ARGS% -cp "%CLASS_PATH%" com.intellij.idea.Main %*

SET PATH=%OLD_PATH%

```


As you can see from the end of it:
`"%JAVA_EXE%" %ALL_JVM_ARGS% -cp "%CLASS_PATH%" com.intellij.idea.Main %*`
it passes ALL_JVM_ARGS first, and user params in the end.
That why if a write smth like `idea.bat  -Didea.exclude.patterns=....` it does not understand that it is JVM arg, but thinks, that is just params to the Main class.

So, I see here 3 solutions:
1) Rewrite `idea.bat` for our needs (copy all logic regarding java location, classpath params, default args, e.t.c.) and provide ` -Didea.exclude.patterns=` flag on top of that.
Cons: It is IDE specific code - could be changed by JB in future versions. And we will have to support it compatility.
2) Do something with `ALL_JVM_ARGS` param. It is partially consists of properties in `USER_VM_OPTIONS_FILE`. So,
  a) We can force user to put  `-Didea.exclude.patterns=` flag in this file.
      Cons: It is not user friendly, and flags in this file affect not only checkstyle project.
  b) We can copy from `idea.bat` logic of finding this file, then dynamicly add `-Didea.exclude.patterns=` flag in the end of this file, run inspections, and finally restore it back.

I'm personnaly think, that 2b) option is the best one. Are you agree?

**Comment by romani** [comment]
@Nimfadora ,

ALL_JVM_ARGS is assembled as
`SET ALL_JVM_ARGS=%ACC% %COMMON_JVM_ARGS% %IDE_JVM_ARGS% `

COMMON_JVM_ARGS as
`SET COMMON_JVM_ARGS=.........%IDE_PROPERTIES_PROPERTY%`

IDE_PROPERTIES_PROPERTY as 
`......%IDEA_PROPERTIES%`

please try to define
`SET IDEA_PROPERTIES="-Didea.exclude.patterns=......"` and launch this bat file.

File: idea_inspection.bat [hunk 2]
```diff
@@ -0,0 +1,37 @@
+@echo off
+
+::----------------------------------------------------------------------
+:: IntelliJ IDEA inspections for checkstyle.
+::----------------------------------------------------------------------
```
Comment (romani) [inline idea_inspection.bat]: Please add example of usage

File: idea_inspection.bat [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+::Try to search in path
+FOR /f "delims=" %%i IN ('"where idea.bat"') DO SET IDEA_LOCATION="%%i"
+if [%IDEA_LOCATION%] NEQ [] (
+    goto run 
+) ELSE (
+    echo IntelliJ IDEA was not found in path
+)
+
+::Ask user for input
```
Comment (romani) [inline idea_inspection.bat]: No interactions please. Fail with error if smth is not defined

File: idea_inspection.bat (hunk 2 above)
Comment (Nimfadora) [reply idea_inspection.bat]: set IDEA_HOME environment variable to be equal to the path to the idea.bat/idea.sh file for ex: 

windows: SET IDEA_HOME=C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2017.2.1\bin\idea.bat 
linux: export IDEA_JOME=/usr/local/bin/idea
mac: export IDEA_HOME="/Applications/IntelliJ IDEA.app/Contents/MacOS/idea"

(directories may be different)

than just run idea-inspection.bat/idea-inspection.sh from checkstyle root dir

File: idea_inspection.bat (hunk 3 above)
Comment (Nimfadora) [reply idea_inspection.bat]: done

File: idea_inspection.bat (hunk 2 above)
Comment (romani) [reply idea_inspection.bat]: just place such examples to bat and shell file. By example user quickly see how to use it.

Please place all that bat/shell files to ./ci folder.

File: idea_inspection.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+PROJECT_DIR=$PWD/
+INSPECTIONS_PATH=$PWD/config/intellij-idea-inspections.xml
+RESULTS_DIR=$PWD/target/inspection-results
+NOISE_LVL=v1
+IDEA_LOCATION=
+export IDEA_PROPERTIES=$PWD/config/idea.properties
+
+# Check IDEA_HOME env variable
+
+if [[ ! -z $IDEA_HOME ]]; then
```
Comment (romani) [inline idea_inspection.sh]: > 
```
$ ./idea_inspection.sh
bash: ./idea_inspection.sh: Permission denied
```

please do `chmod u+x ./idea_inspection.sh` and push to repo file that will be ready to use.

-------------------------
>
 ```
$ ./idea_inspection.sh
./idea_inspection.sh: 16: ./idea_inspection.sh: [[: not found
IDEA_HOME variable not found.
```
first error is not expected.

File: idea_inspection.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#!/bin/sh -e
```
Comment (romani) [inline idea_inspection.sh]: please use bash

File: idea_inspection.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+#################################################
+# IntelliJ IDEA inspections for checkstyle.
+#################################################
```
Comment (romani) [inline idea_inspection.sh]: please add example of usage by single line command.

My try:
```
$ export IDEA_HOME="/home/rivanov/java/idea-IU-171.4249.39/bin/idea.sh" 

$ echo $IDEA_HOME
/home/rivanov/java/idea-IU-171.4249.39/bin/idea.sh

$ ./idea_inspection.sh
./idea_inspection.sh: 16: ./idea_inspection.sh: [[: not found
IDEA_HOME variable not found.
```

File: idea_inspection.bat [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+SET NOISE_LVL=v1
+SET IDEA_LOCATION=
+SET IDEA_PROPERTIES=%CD%\config\idea.properties
+
+::Check IDEA_HOME env variable
+IF EXIST %IDEA_HOME% SET (
+    SET IDEA_LOCATION=%IDEA_HOME%
+    goto run
+) ELSE (
+    echo IDEA_HOME variable not found.
```
Comment (rnveach) [inline idea_inspection.bat]: why don't we exit with failure if this is not set?

File: idea_inspection.bat [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+SET PROJECT_DIR=%CD%\
+SET INSPECTIONS_PATH=%CD%\config\intellij-idea-inspections.xml
+SET RESULTS_DIR=%CD%\target\inspection-results
+SET NOISE_LVL=v1
+SET IDEA_LOCATION=
+SET IDEA_PROPERTIES=%CD%\config\idea.properties
+
+::Check IDEA_HOME env variable
+IF EXIST %IDEA_HOME% SET (
+    SET IDEA_LOCATION=%IDEA_HOME%
```
Comment (rnveach) [inline idea_inspection.bat]: Why can't we use `IDEA_HOME` directly?

Comment (Nimfadora) [reply idea_inspection.bat]: Because next we will try to search idea.bat in path. 
Same as below...

**Comment by Nimfadora** [comment]
@rnveach 
> Can we add support for spaces, or is this a requirement? IDEA installs into "Program Files" by default.

Done.

@romani 

>Please use IDEA_PATH
done

However, something weird with IDEA_PROPERTIES variable. 
Using it I was able to set scope for Mac and Windows. On Linux scope doesn't work... No idea why...

File: idea_inspection.sh (hunk 5 above)
Comment (Nimfadora) [reply idea_inspection.sh]: done (above one also, cannot comment)

File: idea_inspection.sh (hunk 6 above)
Comment (Nimfadora) [reply idea_inspection.sh]: done

File: idea_inspection.sh (hunk 4 above)
Comment (Nimfadora) [reply idea_inspection.sh]: done

**Comment by romani** [comment]
this PR was merged with alot  of fixes for shell version to make it usable.
I will share details of what is left to be done in issue.

**Comment by rnveach** [comment]
> > Can we add support for spaces, or is this a requirement?

> Done.

This still doesn't work for me.
````
M:\checkstyleWorkspace\checkstyle\.ci>set IDEA_PATH=C:\Program Files\JetBrains\I
ntelliJ IDEA Community Edition 2017.2

M:\checkstyleWorkspace\checkstyle\.ci>idea_inspection.bat
The system cannot find the path specified.
'"C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2017.2"' is not rec
ognized as an internal or external command,
operable program or batch file.

M:\checkstyleWorkspace\checkstyle\.ci>set IDEA_PATH="C:\Program Files\JetBrains\
IntelliJ IDEA Community Edition 2017.2"

M:\checkstyleWorkspace\checkstyle\.ci>idea_inspection.bat
Environment variable ( not defined
A subdirectory or file .idea\scopes already exists.
The system cannot find the path specified.
'""C:\Program' is not recognized as an internal or external command,
operable program or batch file.
````

**Comment by rnveach** [comment]
I find the issue. `IDEA_PATH` has to be the full path to the `idea.bat`.
Usually `PATH` means either the HOME or the HOME plus the bin directory.

This is the results of the fixed run:
````
M:\checkstyleWorkspace\checkstyle>M:\checkstyleWorkspace\checkstyle\.ci\idea_ins
pection.bat
A subdirectory or file .idea\scopes already exists.
        1 file(s) copied.

Starting up IntelliJ IDEA 2017.2 (build IC-172.3317.76) ...done.
Opening project...done.
Initializing project...Loaded profile 'Checkstyle' from file 'M:\checkstyleWorks
pace\checkstyle\config\intellij-idea-inspections.xml'
done.
Inspecting with profile 'Checkstyle'
This project contains no modules. Inspection cannot proceed.
````

**Comment by romani** [comment]
> Usually PATH means either the HOME or the HOME plus the bin directory.

It was my request to name it like this. HOME is not good name at all.
Is it better to be `IDEA_BIN` ?

**Comment by romani** [comment]
@rnveach , discussion is moved to https://github.com/checkstyle/checkstyle/issues/4682#issuecomment-326858784

**Pull request closed without merging**